File: podarcisnest/slack/agent.py
"""LLM-powered research agent with scoped access to shared Podarcis knowledge."""

import json
import logging
import re
from pathlib import Path
from typing import Any, Dict, List, Optional

from podarcisnest.slack.config import SlackConfig
from podarcisnest.slack.knowledge import ScopedKnowledgeBase

logger = logging.getLogger("podarcisnest.slack.agent")

SYSTEM_PROMPT = """You are @podarcis, the AI Research Assistant for the PodarcisNest research habitat.
Your primary role is to assist researchers and team members via Slack by:
1. Summarizing recent research progress, newly added protocols, and updated wiki notes.
2. Answering research and scientific questions using the shared Open Knowledge Format (OKF v0.2) wiki.
3. Staging and tracking literature or source papers in the shared repository.

CRITICAL PRIVACY & SECURITY BOUNDARIES:
- You ONLY have access to the team's shared knowledge base (`data/shared/wiki/` and `data/shared/sources/`).
- You DO NOT have access to individual researchers' private workspaces or personal scratchpads.
- Always be accurate, concise, and professional.
- Format responses cleanly for Slack: use emojis, bold headings, bullet points, and reference note paths (e.g. `wiki/protocols/dna_extraction.md`).
"""

TOOLS_SPEC = [
    {
        "name": "get_recent_shared_updates",
        "description": "Retrieve wiki notes and research documents updated or created within the last N days (default 7 days).",
        "parameters": {
            "type": "object",
            "properties": {
                "days": {
                    "type": "integer",
                    "description": "Number of past days to scan for updates (e.g. 7 for last week, 30 for last month).",
                    "default": 7,
                }
            },
        },
    },
    {
        "name": "search_shared_wiki",
        "description": "Search the shared research wiki for notes matching keywords, topics, genes, or protocol names.",
        "parameters": {
            "type": "object",
            "properties": {
                "query": {
                    "type": "string",
                    "description": "Search query or keyword phrase.",
                }
            },
            "required": ["query"],
        },
    },
    {
        "name": "read_shared_note",
        "description": "Read the full text or content of a specific shared note from the wiki or sources.",
        "parameters": {
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "Relative path to note within shared directory (e.g. 'wiki/genomics/sequencing.md' or 'protocols/pcr.md').",
                }
            },
            "required": ["path"],
        },
    },
    {
        "name": "list_shared_sources",
        "description": "List recently added literature papers, datasets, and pending staging queue in data/shared/sources/.",
        "parameters": {
            "type": "object",
            "properties": {},
        },
    },
    {
        "name": "stage_source_url",
        "description": "Stage a scientific paper URL, DOI, or dataset link into the shared queue for team literature ingestion.",
        "parameters": {
            "type": "object",
            "properties": {
                "url": {"type": "string", "description": "URL or DOI of the paper/source."},
                "title": {"type": "string", "description": "Title or short description of the paper."},
                "notes": {"type": "string", "description": "Optional notes or reason for staging."},
            },
            "required": ["url"],
        },
    },
]


class PodarcisResearchAgent:
    """Agent orchestrator that parses user queries, executes scoped tools, and generates responses."""

    def __init__(self, root_dir: Path, config: SlackConfig):
        self.root_dir = root_dir.resolve()
        self.config = config
        self.kb = ScopedKnowledgeBase(self.root_dir)

    def _execute_tool(self, tool_name: str, tool_args: Dict[str, Any], user_name: str = "slack-user") -> Any:
        """Execute scoped knowledge tools."""
        if tool_name == "get_recent_shared_updates":
            days = tool_args.get("days", 7)
            return self.kb.get_recent_wiki_updates(days=days)

        elif tool_name == "search_shared_wiki":
            query = tool_args.get("query", "")
            return self.kb.search_shared_wiki(query=query)

        elif tool_name == "read_shared_note":
            path = tool_args.get("path", "")
            return self.kb.read_shared_note(relative_path=path)

        elif tool_name == "list_shared_sources":
            return self.kb.list_shared_sources()

        elif tool_name == "stage_source_url":
            url = tool_args.get("url", "")
            title = tool_args.get("title")
            notes = tool_args.get("notes", "")
            return self.kb.stage_source_url(url=url, title=title, submitter=user_name, notes=notes)

        return {"error": f"Unknown tool: {tool_name}"}

    def process_message(
        self, user_query: str, user_name: str = "researcher", conversation_history: Optional[List[Dict[str, str]]] = None
    ) -> str:
        """Process an incoming user message using direct tool execution, OpenCode CLI, or direct API."""
        query_lower = user_query.lower().strip()

        # 1. Direct Pattern / Tool Matches (Deterministic, zero external server needed)
        # Summarize recent work / last week
        if any(w in query_lower for w in ["summarize", "summary", "last week", "updates", "recent", "what was done", "what changed"]):
            days = 7
            if "month" in query_lower:
                days = 30
            elif "yesterday" in query_lower or "last 24" in query_lower:
                days = 1
            return self._synthesize_recent_updates(days=days)

        # Staging paper URL / DOI
        if query_lower.startswith("stage ") or "arxiv.org" in query_lower or "doi.org" in query_lower:
            words = user_query.split()
            url = ""
            for w in words:
                if w.startswith("http://") or w.startswith("https://") or "10." in w:
                    url = w
                    break
            if url:
                res = self.kb.stage_source_url(url=url, submitter=user_name, notes=user_query)
                return f"📥 *{res['message']}*\nAdded to `data/shared/sources/staging_queue.json` for team ingestion."

        # List sources / literature queue
        if "sources" in query_lower or "papers" in query_lower or "literature" in query_lower or "queue" in query_lower:
            return self._format_sources_summary()

        # Search shared wiki
        if query_lower.startswith("search ") or query_lower.startswith("find "):
            query_term = re.sub(r"^(search|find)\s+", "", user_query, flags=re.IGNORECASE).strip()
            return self._format_search_results(query_term)

        # 2. OpenCode CLI Execution (Direct subprocess, no server daemon required)
        if self.config.llm_provider == "opencode" and self._has_opencode_cli():
            return self._run_opencode_cli(user_query, user_name)

        # 3. Direct API Call (if API key / compatible endpoint is configured)
        if self.config.llm_api_key and self.config.llm_api_key != "opencode-local":
            if self.config.llm_provider == "anthropic":
                return self._run_anthropic_loop(user_query, user_name, conversation_history)
            else:
                return self._run_openai_loop(user_query, user_name, conversation_history)

        # 4. Fallback search / direct answers if general query
        search_res = self.kb.search_shared_wiki(user_query, max_results=3)
        if search_res:
            return self._format_search_results(user_query)

        # Default help / status summary
        return (
            f"👋 Hi *{user_name}*! I am *@podarcis*, your research assistant.\n\n"
            "Here is what you can ask me directly:\n"
            "• `@podarcis summarize last week` — Generates a progress summary of new and modified OKF notes.\n"
            "• `@podarcis search <topic>` — Searches the shared research wiki for protocols or notes.\n"
            "• `@podarcis list sources` — Shows newly ingested papers and pending staging queue.\n"
            "• `@podarcis stage https://arxiv.org/abs/...` — Stages a paper for literature ingestion."
        )

    def _synthesize_recent_updates(self, days: int = 7) -> str:
        """Directly synthesize recent updates from the shared OKF wiki and sources."""
        recent_notes = self.kb.get_recent_wiki_updates(days=days)
        sources_info = self.kb.list_shared_sources(limit=5)
        staged = sources_info.get("staged_queue", [])

        timeframe_label = "Past 7 Days" if days == 7 else f"Past {days} Days"

        if not recent_notes and not staged:
            return (
                f"📊 *PodarcisNest Shared Research Summary ({timeframe_label})*\n\n"
                f"No modified notes or staged papers found in `data/shared/` over the last {days} days."
            )

        output = [f"📊 *PodarcisNest Shared Research Summary ({timeframe_label})*\n"]

        if recent_notes:
            output.append(f"*📝 Updated OKF Wiki Notes ({len(recent_notes)}):*")
            for note in recent_notes[:8]:
                tags_str = f" `[{', '.join(note['tags'])}]`" if note.get("tags") else ""
                summary_str = f"\n  _{note['summary']}_" if note.get("summary") else ""
                output.append(f"• *{note['title']}* (`{note['path']}`){tags_str}{summary_str}")
            if len(recent_notes) > 8:
                output.append(f"_...and {len(recent_notes) - 8} more notes in `data/shared/wiki/`._")
            output.append("")

        if staged:
            output.append(f"*📚 Pending Literature Staging Queue ({len(staged)}):*")
            for item in staged[:5]:
                output.append(f"• *{item.get('title', item.get('url'))}* (submitted by _{item.get('submitted_by', 'team')}_)")
            output.append("")

        return "\n".join(output).strip()

    def _format_search_results(self, query: str) -> str:
        """Format search results directly from the shared wiki."""
        results = self.kb.search_shared_wiki(query, max_results=5)
        if not results:
            return f"🔍 No shared wiki notes found matching: *{query}* in `data/shared/wiki/`."

        output = [f"🔍 *Shared Wiki Search Results for \"{query}\":*\n"]
        for r in results:
            output.append(f"• *{r['title']}* (`{r['path']}`)")
            if r.get("excerpt"):
                output.append(f"  _{r['excerpt']}_")
        return "\n".join(output)

    def _format_sources_summary(self) -> str:
        """Format sources and staging queue summary."""
        info = self.kb.list_shared_sources(limit=10)
        sources = info.get("sources", [])
        staged = info.get("staged_queue", [])

        output = ["📚 *Shared Literature & Sources Library (`data/shared/sources/`):*\n"]
        if staged:
            output.append(f"*Pending Ingestion Queue ({len(staged)}):*")
            for s in staged:
                output.append(f"• `{s.get('url')}` — *{s.get('title')}* (by _{s.get('submitted_by')}_)")
            output.append("")

        if sources:
            output.append(f"*Available Files & Datasets ({len(sources)}):*")
            for s in sources:
                kind = "📁 Directory" if s["is_dir"] else f"📄 File ({round((s['size_bytes'] or 0)/1024, 1)} KB)"
                output.append(f"• `{s['name']}` — {kind}")
        else:
            output.append("No literature files in `data/shared/sources/` yet.")

        return "\n".join(output)

    def _has_opencode_cli(self) -> bool:
        """Check if opencode CLI is available in PATH."""
        import shutil
        return shutil.which("opencode") is not None

    def _run_opencode_cli(self, user_query: str, user_name: str) -> str:
        """Run OpenCode CLI directly against the shared workspace."""
        import subprocess

        cmd = [
            "opencode",
            "run",
            f"You are @podarcis in Slack responding to {user_name}. Query: {user_query}. You only have access to shared/wiki and shared/sources.",
        ]

        try:
            res = subprocess.run(
                cmd,
                cwd=str(self.kb.shared_dir),
                capture_output=True,
                text=True,
                timeout=60,
            )
            if res.returncode == 0 and res.stdout.strip():
                return res.stdout.strip()
            elif res.stderr:
                return f"⚠️ OpenCode execution returned: {res.stderr.strip()}"
        except Exception as e:
            logger.warning(f"Failed to run opencode CLI: {e}")

        # Fallback to direct synthesis if CLI times out or fails
        return self._synthesize_recent_updates(days=7)

    def _run_anthropic_loop(
        self, user_query: str, user_name: str, history: Optional[List[Dict[str, str]]] = None
    ) -> str:
        """Execute Anthropic Claude agent loop with tool calling."""
        try:
            import anthropic
        except ImportError:
            return "❌ `anthropic` Python package is not installed. Run `pip install anthropic`."

        client = anthropic.Anthropic(api_key=self.config.llm_api_key)
        model_name = self.config.llm_model or "claude-3-5-sonnet-20241022"

        anthropic_tools = []
        for t in TOOLS_SPEC:
            anthropic_tools.append(
                {
                    "name": t["name"],
                    "description": t["description"],
                    "input_schema": t["parameters"],
                }
            )

        messages = []
        if history:
            for h in history:
                messages.append({"role": h["role"], "content": h["content"]})
        messages.append({"role": "user", "content": f"User ({user_name}): {user_query}"})

        # Run up to 5 tool-calling turns
        for _ in range(5):
            response = client.messages.create(
                model=model_name,
                max_tokens=2048,
                system=SYSTEM_PROMPT,
                tools=anthropic_tools,
                messages=messages,
            )

            # Check if tools were requested
            tool_calls = [c for c in response.content if c.type == "tool_use"]
            if not tool_calls:
                text_blocks = [c.text for c in response.content if c.type == "text"]
                return "\n".join(text_blocks).strip()

            # Append assistant message with tool calls
            messages.append({"role": "assistant", "content": response.content})

            # Execute tool calls
            tool_results = []
            for tool_call in tool_calls:
                result = self._execute_tool(tool_call.name, tool_call.input, user_name=user_name)
                tool_results.append(
                    {
                        "type": "tool_result",
                        "tool_use_id": tool_call.id,
                        "content": json.dumps(result),
                    }
                )

            messages.append({"role": "user", "content": tool_results})

        return "I completed the queries but reached the maximum reasoning iterations."

    def _run_openai_loop(
        self, user_query: str, user_name: str, history: Optional[List[Dict[str, str]]] = None
    ) -> str:
        """Execute OpenCode / OpenAI-compatible agent loop with tool calling."""
        try:
            import openai
        except ImportError:
            return "❌ `openai` Python package is not installed. Run `pip install openai`."

        client_kwargs = {"api_key": self.config.llm_api_key or "opencode-local"}
        if self.config.llm_base_url:
            client_kwargs["base_url"] = self.config.llm_base_url

        client = openai.OpenAI(**client_kwargs)
        model_name = self.config.llm_model or "opencode"

        openai_tools = [{"type": "function", "function": t} for t in TOOLS_SPEC]

        messages = [{"role": "system", "content": SYSTEM_PROMPT}]
        if history:
            for h in history:
                messages.append({"role": h["role"], "content": h["content"]})
        messages.append({"role": "user", "content": f"User ({user_name}): {user_query}"})

        for _ in range(5):
            response = client.chat.completions.create(
                model=model_name,
                messages=messages,
                tools=openai_tools,
            )
            msg = response.choices[0].message

            if not msg.tool_calls:
                return msg.content or ""

            messages.append(msg)

            for tc in msg.tool_calls:
                fn_name = tc.function.name
                try:
                    fn_args = json.loads(tc.function.arguments)
                except Exception:
                    fn_args = {}
                result = self._execute_tool(fn_name, fn_args, user_name=user_name)
                messages.append(
                    {
                        "role": "tool",
                        "tool_call_id": tc.id,
                        "content": json.dumps(result),
                    }
                )

        return "I completed the queries but reached the maximum reasoning iterations."

File: src/slack/agent.ts
import path from 'path';
import { spawnSync } from 'child_process';
import Anthropic from '@anthropic-ai/sdk';
import OpenAI from 'openai';
import { SlackConfig } from './config.js';
import { ScopedKnowledgeBase } from './knowledge.js';

export const SYSTEM_PROMPT = `You are @podarcis, the AI Research Assistant for the PodarcisNest research habitat.
Your primary role is to assist researchers and team members via Slack by:
1. Summarizing recent research progress, newly added protocols, and updated wiki notes.
2. Answering research and scientific questions using the shared Open Knowledge Format (OKF v0.2) wiki.
3. Staging and tracking literature or source papers in the shared repository.

CRITICAL PRIVACY & SECURITY BOUNDARIES:
- You ONLY have access to the team's shared knowledge base (\`data/shared/wiki/\` and \`data/shared/sources/\`).
- You DO NOT have access to individual researchers' private workspaces or personal scratchpads.
- Always be accurate, concise, and professional.
- Format responses cleanly for Slack: use emojis, bold headings, bullet points, and reference note paths (e.g. \`wiki/protocols/dna_extraction.md\`).
`;

export const TOOLS_SPEC = [
  {
    name: 'get_recent_shared_updates',
    description: 'Retrieve wiki notes and research documents updated or created within the last N days (default 7 days).',
    parameters: {
      type: 'object',
      properties: {
        days: {
          type: 'integer',
          description: 'Number of past days to scan for updates (e.g. 7 for last week, 30 for last month).',
          default: 7,
        },
      },
    },
  },
  {
    name: 'search_shared_wiki',
    description: 'Search the shared research wiki for notes matching keywords, topics, genes, or protocol names.',
    parameters: {
      type: 'object',
      properties: {
        query: {
          type: 'string',
          description: 'Search query or keyword phrase.',
        },
      },
      required: ['query'],
    },
  },
  {
    name: 'read_shared_note',
    description: 'Read the full text or content of a specific shared note from the wiki or sources.',
    parameters: {
      type: 'object',
      properties: {
        path: {
          type: 'string',
          description: "Relative path to note within shared directory (e.g. 'wiki/genomics/sequencing.md' or 'protocols/pcr.md').",
        },
      },
      required: ['path'],
    },
  },
  {
    name: 'list_shared_sources',
    description: 'List recently added literature papers, datasets, and pending staging queue in data/shared/sources/.',
    parameters: {
      type: 'object',
      properties: {},
    },
  },
  {
    name: 'stage_source_url',
    description: 'Stage a scientific paper URL, DOI, or dataset link into the shared queue for team literature ingestion.',
    parameters: {
      type: 'object',
      properties: {
        url: { type: 'string', description: 'URL or DOI of the paper/source.' },
        title: { type: 'string', description: 'Title or short description of the paper.' },
        notes: { type: 'string', description: 'Optional notes or reason for staging.' },
      },
      required: ['url'],
    },
  },
];

export class PodarcisResearchAgent {
  public rootDir: string;
  public config: SlackConfig;
  public kb: ScopedKnowledgeBase;

  constructor(rootDir: string, config: SlackConfig) {
    this.rootDir = path.resolve(rootDir);
    this.config = config;
    this.kb = new ScopedKnowledgeBase(this.rootDir);
  }

  public executeTool(toolName: string, toolArgs: Record<string, any>, userName: string = 'slack-user'): any {
    switch (toolName) {
      case 'get_recent_shared_updates':
        return this.kb.getRecentWikiUpdates(toolArgs.days ?? 7);
      case 'search_shared_wiki':
        return this.kb.searchSharedWiki(toolArgs.query || '');
      case 'read_shared_note':
        return this.kb.readSharedNote(toolArgs.path || '');
      case 'list_shared_sources':
        return this.kb.listSharedSources();
      case 'stage_source_url':
        return this.kb.stageSourceUrl(toolArgs.url || '', toolArgs.title, userName, toolArgs.notes || '');
      default:
        return { error: `Unknown tool: ${toolName}` };
    }
  }

  public async processMessage(
    userQuery: string,
    userName: string = 'researcher',
    conversationHistory?: Array<{ role: 'user' | 'assistant'; content: string }>
  ): Promise<string> {
    const queryLower = userQuery.toLowerCase().trim();

    // 1. Direct Pattern Matches
    if (['summarize', 'summary', 'last week', 'updates', 'recent', 'what was done', 'what changed'].some((w) => queryLower.includes(w))) {
      let days = 7;
      if (queryLower.includes('month')) days = 30;
      else if (queryLower.includes('yesterday') || queryLower.includes('last 24')) days = 1;
      return this.synthesizeRecentUpdates(days);
    }

    if (queryLower.startsWith('research ') || queryLower.startsWith('@researcher ')) {
      const searchQuery = userQuery.replace(/^(@researcher|research)\s+/i, '').trim();
      return (
        `🔬 *Spawning @researcher for:* \`${searchQuery}\`\n\n` +
        `Searching academic sources (Semantic Scholar / PubMed) and enqueuing results into \`data/shared/sources/\`.\n` +
        `Use \`@podarcis list sources\` to view enqueued literature once complete.`
      );
    }

    if (['synthesize', 'ingest', '@synthesizer', 'run synthesis'].includes(queryLower)) {
      const sourcesInfo = this.kb.listSharedSources(5);
      const staged = sourcesInfo.staged_queue;
      return (
        `🧠 *Spawning @synthesizer on Shared Knowledge Base:*\n\n` +
        `Processing ${staged.length} pending source(s) from \`data/shared/sources/staging_queue.json\` into OKF v0.2 wiki notes in \`data/shared/wiki/\`.\n` +
        `Running \`@auditor\` machine verification loop afterwards.`
      );
    }

    if (queryLower.startsWith('audit') || queryLower.startsWith('@auditor')) {
      return (
        `🔍 *Spawning @auditor Machine Verification:*\n\n` +
        `Running link integrity check, citation validation, and OKF v0.2 schema linting on \`data/shared/wiki/\`.\n` +
        `Verified notes will be signed off with status \`stable\`.`
      );
    }

    if (queryLower.startsWith('stage ') || queryLower.includes('arxiv.org') || queryLower.includes('doi.org')) {
      const words = userQuery.split(/\s+/);
      const url = words.find((w) => w.startsWith('http://') || w.startsWith('https://') || w.includes('10.'));
      if (url) {
        const res = this.kb.stageSourceUrl(url, undefined, userName, userQuery);
        return `📥 *${res.message}*\nAdded to \`data/shared/sources/staging_queue.json\` for team ingestion.`;
      }
    }

    if (['sources', 'papers', 'literature', 'queue'].some((w) => queryLower.includes(w))) {
      return this.formatSourcesSummary();
    }

    if (queryLower.startsWith('search ') || queryLower.startsWith('find ')) {
      const queryTerm = userQuery.replace(/^(search|find)\s+/i, '').trim();
      return this.formatSearchResults(queryTerm);
    }

    // 2. OpenCode CLI Execution
    if (this.config.llm_provider === 'opencode' && this.hasOpencodeCli()) {
      return this.runOpencodeCli(userQuery, userName);
    }

    // 3. Direct LLM API Execution
    if (this.config.llm_api_key && this.config.llm_api_key !== 'opencode-local') {
      if (this.config.llm_provider === 'anthropic') {
        return this.runAnthropicLoop(userQuery, userName, conversationHistory);
      } else {
        return this.runOpenAiLoop(userQuery, userName, conversationHistory);
      }
    }

    // 4. Fallback search
    const searchRes = this.kb.searchSharedWiki(userQuery, 3);
    if (searchRes.length > 0) {
      return this.formatSearchResults(userQuery);
    }

    // Default Help
    return (
      `👋 Hi *${userName}*! I am *@podarcis*, your research team assistant.\n\n` +
      `Here is what you can ask me directly in Slack:\n` +
      `• \`@podarcis research <query>\` — Discovers literature and downloads papers via \`@researcher\`\n` +
      `• \`@podarcis synthesize\` — Compiles pending sources into OKF wiki notes via \`@synthesizer\`\n` +
      `• \`@podarcis audit\` — Runs machine verification and citation audits via \`@auditor\`\n` +
      `• \`@podarcis summarize last week\` — Progress summary of new and modified OKF notes\n` +
      `• \`@podarcis search <topic>\` — Searches the shared research wiki for protocols or notes\n` +
      `• \`@podarcis stage https://arxiv.org/abs/...\` — Stages a paper for literature ingestion\n` +
      `• \`@podarcis list sources\` — Shows newly ingested papers and pending staging queue`
    );
  }

  private synthesizeRecentUpdates(days: number = 7): string {
    const recentNotes = this.kb.getRecentWikiUpdates(days);
    const sourcesInfo = this.kb.listSharedSources(5);
    const staged = sourcesInfo.staged_queue;

    const timeframeLabel = days === 7 ? 'Past 7 Days' : `Past ${days} Days`;

    if (recentNotes.length === 0 && staged.length === 0) {
      return (
        `📊 *PodarcisNest Shared Research Summary (${timeframeLabel})*\n\n` +
        `No modified notes or staged papers found in \`data/shared/\` over the last ${days} days.`
      );
    }

    const output = [`📊 *PodarcisNest Shared Research Summary (${timeframeLabel})*\n`];

    if (recentNotes.length > 0) {
      output.push(`*📝 Updated OKF Wiki Notes (${recentNotes.length}):*`);
      for (const note of recentNotes.slice(0, 8)) {
        const tagsStr = note.tags && note.tags.length > 0 ? ` \`[${note.tags.join(', ')}]\`` : '';
        const summaryStr = note.summary ? `\n  _${note.summary}_` : '';
        output.push(`• *${note.title}* (\`${note.path}\`)${tagsStr}${summaryStr}`);
      }
      if (recentNotes.length > 8) {
        output.push(`_...and ${recentNotes.length - 8} more notes in \`data/shared/wiki/\`._`);
      }
      output.push('');
    }

    if (staged.length > 0) {
      output.push(`*📚 Pending Literature Staging Queue (${staged.length}):*`);
      for (const item of staged.slice(0, 5)) {
        output.push(`• *${item.title || item.url}* (submitted by _${item.submitted_by || 'team'}_)`);
      }
      output.push('');
    }

    return output.join('\n').trim();
  }

  private formatSearchResults(query: string): string {
    const results = this.kb.searchSharedWiki(query, 5);
    if (results.length === 0) {
      return `🔍 No shared wiki notes found matching: *${query}* in \`data/shared/wiki/\`.`;
    }

    const output = [`🔍 *Shared Wiki Search Results for "${query}":*\n`];
    for (const r of results) {
      output.push(`• *${r.title}* (\`${r.path}\`)`);
      if (r.excerpt) {
        output.push(`  _${r.excerpt}_`);
      }
    }
    return output.join('\n');
  }

  private formatSourcesSummary(): string {
    const info = this.kb.listSharedSources(10);
    const sources = info.sources;
    const staged = info.staged_queue;

    const output = ['📚 *Shared Literature & Sources Library (`data/shared/sources/`):*\n'];
    if (staged.length > 0) {
      output.push(`*Pending Ingestion Queue (${staged.length}):*`);
      for (const s of staged) {
        output.push(`• \`${s.url}\` — *${s.title}* (by _${s.submitted_by}_)`);
      }
      output.push('');
    }

    if (sources.length > 0) {
      output.push(`*Available Files & Datasets (${sources.length}):*`);
      for (const s of sources) {
        const kind = s.is_dir ? '📁 Directory' : `📄 File (${Math.round((s.size_bytes || 0) / 1024)} KB)`;
        output.push(`• \`${s.name}\` — ${kind}`);
      }
    } else {
      output.push('No literature files in `data/shared/sources/` yet.');
    }

    return output.join('\n');
  }

  private hasOpencodeCli(): boolean {
    const res = spawnSync('which', ['opencode'], { stdio: 'pipe' });
    return res.status === 0;
  }

  private runOpencodeCli(userQuery: string, userName: string): string {
    const prompt = `You are @podarcis in Slack responding to ${userName}. Query: ${userQuery}. You only have access to shared/wiki and shared/sources.`;
    const res = spawnSync('opencode', ['run', prompt], {
      cwd: this.kb.sharedDir,
      encoding: 'utf-8',
      timeout: 60000,
    });
    if (res.status === 0 && res.stdout && res.stdout.trim()) {
      return res.stdout.trim();
    }
    return this.synthesizeRecentUpdates(7);
  }

  private async runAnthropicLoop(
    userQuery: string,
    userName: string,
    history?: Array<{ role: 'user' | 'assistant'; content: string }>
  ): Promise<string> {
    const client = new Anthropic({ apiKey: this.config.llm_api_key });
    const model = this.config.llm_model || 'claude-3-5-sonnet-20241022';

    const tools: Anthropic.Tool[] = TOOLS_SPEC.map((t) => ({
      name: t.name,
      description: t.description,
      input_schema: t.parameters as Anthropic.Tool.InputSchema,
    }));

    const messages: Anthropic.MessageParam[] = [];
    if (history) {
      for (const h of history) {
        messages.push({ role: h.role, content: h.content });
      }
    }
    messages.push({ role: 'user', content: `User (${userName}): ${userQuery}` });

    for (let turn = 0; turn < 5; turn++) {
      const response = await client.messages.create({
        model,
        max_tokens: 2048,
        system: SYSTEM_PROMPT,
        tools,
        messages,
      });

      const toolCalls = response.content.filter((c) => c.type === 'tool_use') as Anthropic.ToolUseBlock[];
      if (toolCalls.length === 0) {
        const textBlocks = response.content.filter((c) => c.type === 'text') as Anthropic.TextBlock[];
        return textBlocks.map((t) => t.text).join('\n').trim();
      }

      messages.push({ role: 'assistant', content: response.content });

      const toolResults: Anthropic.ToolResultBlockParam[] = [];
      for (const tc of toolCalls) {
        const res = this.executeTool(tc.name, tc.input as Record<string, any>, userName);
        toolResults.push({
          type: 'tool_result',
          tool_use_id: tc.id,
          content: JSON.stringify(res),
        });
      }

      messages.push({ role: 'user', content: toolResults });
    }

    return 'I completed the queries but reached the maximum reasoning iterations.';
  }

  private async runOpenAiLoop(
    userQuery: string,
    userName: string,
    history?: Array<{ role: 'user' | 'assistant'; content: string }>
  ): Promise<string> {
    const client = new OpenAI({
      apiKey: this.config.llm_api_key || 'opencode-local',
      baseURL: this.config.llm_base_url,
    });
    const model = this.config.llm_model || 'opencode';

    const tools: OpenAI.Chat.ChatCompletionTool[] = TOOLS_SPEC.map((t) => ({
      type: 'function',
      function: {
        name: t.name,
        description: t.description,
        parameters: t.parameters,
      },
    }));

    const messages: OpenAI.Chat.ChatCompletionMessageParam[] = [
      { role: 'system', content: SYSTEM_PROMPT },
    ];
    if (history) {
      for (const h of history) {
        messages.push({ role: h.role, content: h.content });
      }
    }
    messages.push({ role: 'user', content: `User (${userName}): ${userQuery}` });

    for (let turn = 0; turn < 5; turn++) {
      const response = await client.chat.completions.create({
        model,
        messages,
        tools,
      });

      const msg = response.choices[0].message;
      if (!msg.tool_calls || msg.tool_calls.length === 0) {
        return msg.content || '';
      }

      messages.push(msg);

      for (const tc of msg.tool_calls) {
        let args = {};
        try {
          args = JSON.parse(tc.function.arguments);
        } catch {}
        const res = this.executeTool(tc.function.name, args, userName);
        messages.push({
          role: 'tool',
          tool_call_id: tc.id,
          content: JSON.stringify(res),
        });
      }
    }

    return 'I completed the queries but reached the maximum reasoning iterations.';
  }
}
